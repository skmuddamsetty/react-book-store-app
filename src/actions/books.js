// import uuid from 'uuid';
import database from '../firebase/firebase';

// ADD_BOOK
export const addBook = book => ({
  type: 'ADD_BOOK',
  book
});
// REMOVE_BOOK
export const removeBook = ({ id } = {}) => ({
  type: 'REMOVE_BOOK',
  id
});
// EDIT_BOOK

export const editBook = (id, updates) => ({
  type: 'EDIT_BOOK',
  id,
  updates
});

export const startAddBook = (bookData = {}) => {
  return dispatch => {
    const { description = '', note = '', price = 0, createdAt = 0 } = bookData;
    const book = { description, note, price, createdAt };
    return database
      .ref('books')
      .push(book)
      .then(ref => {
        dispatch(addBook({ id: ref.key, ...book }));
      })
      .catch();
  };
};
