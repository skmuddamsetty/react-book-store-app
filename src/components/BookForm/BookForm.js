import React, { Component } from 'react';
class BookForm extends Component {
  state = { description: '', note: '', price: '' };

  descriptionChangeHandler = e => {
    const description = e.target.value;
    this.setState(() => ({
      description
    }));
  };

  noteChangeHandler = e => {
    const note = e.target.value;
    this.setState(() => ({
      note
    }));
  };

  priceChangeHandler = e => {
    const price = e.target.value;
    if (price.match(/^\d*(\.\d{0,2})?$/)) {
      this.setState(() => ({ price }));
    }
  };

  render() {
    return (
      <div>
        <form>
          <input
            type='text'
            placeholder='Description'
            autoFocus
            value={this.state.description}
            onChange={this.descriptionChangeHandler}
          />
          <input
            type='text'
            placeholder='Price'
            value={this.state.price}
            onChange={this.priceChangeHandler}
          />
          <textarea
            placeholder='Add note for your book (optional)'
            onChange={this.noteChangeHandler}
            value={this.state.note}
          ></textarea>
          <button type='submit'>Add Book</button>
        </form>
      </div>
    );
  }
}

export default BookForm;
