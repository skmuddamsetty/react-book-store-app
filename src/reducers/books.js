// Books reducer
const booksReducerDefaultState = [];

export const booksReducer = (state = booksReducerDefaultState, action) => {
  switch (action.type) {
    case 'ADD_BOOK':
      //   return state.concat(action.book);
      return [...state, action.book];
    case 'REMOVE_BOOK':
      return state.filter(book => book.id !== action.id);
    case 'EDIT_BOOK':
      return state.map(book => {
        if (book.id === action.id) {
          return { ...book, ...action.updates };
        } else {
          return book;
        }
      });
    default:
      return state;
  }
};
