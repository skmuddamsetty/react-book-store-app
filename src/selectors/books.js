// get Visible Books
export const getVisibleBooks = (
  books,
  { text, sortBy, startDate, endDate }
) => {
  return books
    .filter(book => {
      const startDateMatch =
        typeof startDate !== 'number' || book.createdAt >= startDate;
      const endDateMatch =
        typeof endDate !== 'number' || books.createdAt <= endDate;
      const textMatch = book.description
        .toLowerCase()
        .includes(text.toLowerCase());
      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === 'date') {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === 'price') {
        return a.price < b.price ? -1 : 1;
      } else {
        return 0;
      }
    });
};
