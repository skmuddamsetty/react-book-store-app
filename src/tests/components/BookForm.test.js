import React from 'react';
import { shallow } from 'enzyme';
import { books } from '../fixtures/books';
import toJSON from 'enzyme-to-json';
import BookForm from '../../components/BookForm/BookForm';

test('should render book form properly', () => {
  const wrapper = shallow(<BookForm />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render book form with book data', () => {
  const wrapper = shallow(<BookForm book={books[1]} />);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<BookForm />);
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(toJSON(wrapper)).toMatchSnapshot();
});

test('should set description on input change', () => {
  const value = 'New Description';
  const wrapper = shallow(<BookForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', { target: { value } });
  expect(wrapper.state('description')).toBe(value);
});

test('should set note on input change', () => {
  const value = 'New Note';
  const wrapper = shallow(<BookForm />);
  wrapper.find('textarea').simulate('change', { target: { value } });
  expect(wrapper.state('note')).toBe(value);
});

test('should set price if valid price', () => {
  const value = '12.22';
  const wrapper = shallow(<BookForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', { target: { value } });
  expect(wrapper.state('price')).toBe(value);
});

test('should not set price if invalid price', () => {
  const value = '12.222';
  const wrapper = shallow(<BookForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', { target: { value } });
  expect(wrapper.state('price')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(<BookForm book={books[0]} onSubmit={onSubmitSpy} />);
  wrapper.find('form').simulate('submit', { preventDefault: () => {} });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: books[0].description,
    price: books[0].price,
    note: books[0].note
  });
});
