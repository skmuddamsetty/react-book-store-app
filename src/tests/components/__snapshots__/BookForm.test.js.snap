// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render book form properly 1`] = `
<div
  className="container"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        for="description"
      >
        Description
      </label>
      <input
        autoFocus={true}
        className="form-control"
        id="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="price"
      >
        Price
      </label>
      <input
        className="form-control"
        id="price"
        onChange={[Function]}
        placeholder="Price"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="note"
      >
        Note
      </label>
      <textarea
        className="form-control"
        id="note"
        onChange={[Function]}
        placeholder="Add note for your book (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-success"
      type="submit"
    >
      Add Book
    </button>
  </form>
</div>
`;

exports[`should render book form with book data 1`] = `
<div
  className="container"
>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        for="description"
      >
        Description
      </label>
      <input
        autoFocus={true}
        className="form-control"
        id="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        type="text"
        value="Angular Book"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="price"
      >
        Price
      </label>
      <input
        className="form-control"
        id="price"
        onChange={[Function]}
        placeholder="Price"
        type="text"
        value="0.1"
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="note"
      >
        Note
      </label>
      <textarea
        className="form-control"
        id="note"
        onChange={[Function]}
        placeholder="Add note for your book (optional)"
        value="Angular Book Notes"
      />
    </div>
    <button
      className="btn btn-success"
      type="submit"
    >
      Add Book
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div
  className="container"
>
  <p>
    Please provider Description and Price
  </p>
  <form
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <label
        for="description"
      >
        Description
      </label>
      <input
        autoFocus={true}
        className="form-control"
        id="description"
        onChange={[Function]}
        placeholder="Description"
        required={true}
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="price"
      >
        Price
      </label>
      <input
        className="form-control"
        id="price"
        onChange={[Function]}
        placeholder="Price"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <label
        for="note"
      >
        Note
      </label>
      <textarea
        className="form-control"
        id="note"
        onChange={[Function]}
        placeholder="Add note for your book (optional)"
        value=""
      />
    </div>
    <button
      className="btn btn-success"
      type="submit"
    >
      Add Book
    </button>
  </form>
</div>
`;
